src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcProxyShimFactory.cs

```diff
@@ -47,7 +47,30 @@ internal DtcProxyShimFactory(EventWaitHandle notificationEventHandle)
         object? pvConfigPararms,
         [MarshalAs(UnmanagedType.Interface)] out ITransactionDispenser ppvObject);
 
-    [UnconditionalSuppressMessage("Trimming", "IL2050", Justification = "Leave me alone")]
+    [RequiresUnreferencedCode("Distributed transactions support is not compatible with trimming. Correctness of the application cannot be guaranteed after trimming.")]
+    [DynamicDependency(DynamicallyAccessedMemberTypes.All, typeof(ITransactionDispenser))]
```

Comment:
Is it worth leaving all these in, given that they're insufficient to make trimming work?

Reply:
My thinking was that if someone was actually trying to use this functionality in a trimmed app, and they were trying to make it work, they wouldn't need to do this manually themselves. Yes, it isn't guaranteed to work, but we can at least reduce the steps someone would need to do to make it work if they really wanted to.

Reply:
The thing is that unless I'm mistaken, a trimmed app won't work because of the COM interop, and I'm not sure it's possible for a user to make it work (because of https://github.com/dotnet/runtime/pull/75176#discussion_r964340209) (so it's not about guaranteeing or not - I think we know it won't work). If that's indeed the case, then these may provide some false impression that things work (or can be made to work) when that's not the case.

But am OK merging these in any case.

Reply:
Something like the following should trigger this code path:

```c#
using (var scope = new TransactionScope())
using (var conn1 = new SqlConnection(@"Server=(localdb)\MSSQLLocalDB;Integrated Security=true"))
using (var conn2 = new SqlConnection(@"Server=(localdb)\MSSQLLocalDB;Integrated Security=true"))
{
    conn1.Open();
    conn2.Open();
    
    Console.WriteLine("Escalated to distributed transaction. Press enter to commit.");
    Console.ReadLine();
    
    scope.Complete();
}
```

Note that it's definitely a minimal app that doesn't exercise all the distributed transaction code paths, but if this works it's definitely a start.

Reply:
Thanks! This helped me test the code as well and caught that I wasn't using all the right DynamicDependencies. I've added a couple more to the interfaces I missed. With that change, I can now publish and successfully run this app:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <PublishTrimmed>true</PublishTrimmed>
    <RuntimeIdentifier>win-x64</RuntimeIdentifier>
    <BuiltInComInteropSupport>true</BuiltInComInteropSupport>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Data.SqlClient" Version="5.0.0" />
  </ItemGroup>

</Project>
```

```C#
using Microsoft.Data.SqlClient;
using System.Transactions;

using (var scope = new TransactionScope())
using (var conn1 = new SqlConnection(@"Server=(localdb)\MSSQLLocalDB;Integrated Security=true"))
using (var conn2 = new SqlConnection(@"Server=(localdb)\MSSQLLocalDB;Integrated Security=true"))
{
    conn1.Open();
    conn2.Open();

    Console.WriteLine("Escalated to distributed transaction. Press enter to commit.");
    Console.ReadLine();
    Console.WriteLine("Completing.");

    scope.Complete();
}

Console.WriteLine("Done.");
```

The only thing someone would need to do is set `<BuiltInComInteropSupport>true</BuiltInComInteropSupport>` on top of `PublishTrimmed=true`.

Reply:
Great! Do we think this "guarantees" that it works? :)

I think the following interfaces are still missing, for other non-basic flows ([the list of COM interfaces is here](https://github.com/dotnet/runtime/tree/main/src/libraries/System.Transactions.Local/src/System/Transactions/DtcProxyShim/DtcInterfaces)):

* IPrepareInfo
* ITransactionCloner
* ITransactionPhase0EnlistmentAsync
* ITransactionPhase0Factory
* ITransactionPhase0NotifyAsync

Assuming we still want the RequiresUnreferencedCode just in case, we may want to change the message to something like "Distributed transactions support may not be compatible with trimming. If your program creates a distributed transaction via System.Transactions, the correctness of the application cannot be guaranteed after trimming.")]

Thanks again for spending time on this!

/cc @AaronRobinsonMSFT

Reply:
> Great! Do we think this "guarantees" that it works? :)

No. There isn't a guarantee here until someone rewrites all of this to use COM Wrappers. That's why the warning is in place. I also didn't test it with NativeAOT, which I don't believe would work (probably at all, even with the BuiltInComInteropSupport=true).

> I think the following interfaces are still missing

I can add those and sort the list.

> we may want to change the message to something like

I will update the message.

Reply:
Thanks, sounds good!

Reply:
I'll leave this up to Aaron because any bugs with this will inevitably land on his plate, but I'm a bit uneasy adding extra annotations so that the very basic scenario works and leaving everything else untested because there's a warning and things are allowed not to work.

I don't like when we generate useless warnings because it desensitizes users to them.

I would lean towards either having a warning and doing _nothing_ to make things works with trimming, or ensuring this works and declare it a supported scenario with no warnings. I would lean towards the former.

Reply:
> ensuring this works and declare it a supported scenario with no warnings

Since the only way to do that is to convert all this to COMWrappers, this would take a lot more work than worth it at this point. So I'd say that option is off the table in the near term.

> I don't like when we generate useless warnings because it desensitizes users to them.

The scenario still doesn't work by default with trimming. The user still needs to manually enable BuiltInComInteropSupport before it will work. So it isn't useless.

> having a warning and doing nothing to make things works with trimming

In the end, if a customer is actually trying to make this work, I think these annotations help them. They don't need to find this list themselves and somehow figure out how to reference all these internal types in either their own DynamicDependency attributes or in a ILLink.Descriptors.xml file.

Reply:
> The scenario still doesn't work by default with trimming. The user still needs to manually enable BuiltInComInteropSupport before it will work. So it isn't useless.

The message in the RUC attribute doesn't mention COM.

> In the end, if a customer is actually trying to make this work, I think these annotations help them.

They help them on the golden path that was tested. I don't know how wide that path is.

In general it feels better when unsupported things crash the moment `PublishTrimmed` is enabled on a project (or S.T.Local references are added) than if they appear to work and hit the _very obscure_ COM failure modes long after `PublishTrimmed` was enabled on the project and people already forgot about it and got desensitized to the warning. If we don't try to help them, most people will give up on trimming with this assembly and that honestly feels like a better outcome.

Reply:
> The message in the RUC attribute doesn't mention COM.

That was intentional. I left it vague because like you said there may be other obscure holes waiting around the corner. I didn't want to mention "to make it work, enable COM".

> most people will give up on trimming with this assembly and that honestly feels like a better outcome.

I think this the root of where we disagree. If someone needs to trim their app, I'd rather get them part of the way there, if possible/reasonable. There's a reason they are trying to do it. It doesn't seem like a good customer experience to not do these simple things to help them.

Reply:
> I think this the root of where we disagree. If someone needs to trim their app, I'd rather get them part of the way there, if possible/reasonable

I usually have the same position, but draw a very strong line for COM. The failure modes are undiagnosable for a pure .NET developer.

Reply:
Since I really want the other changes in this PR in, I've removed the DynamicDependency attributes. They are lower importance than:

1. Getting the developer a warning when using this code in trimmed/NativeAOT apps
2. Removing the size regression on Windows when using `partial` trimming